Algorithms/Search/KmpSearcher.cs

```diff
@@ -0,0 +1,101 @@
+﻿using System.Collections.Generic;
+using Utilities.Exceptions;
+
+namespace Algorithms.Search
+{
+    /// <summary>
+    /// An implementation of Knuth–Morris–Pratt Algorithm.
+    /// Worst case time complexity: O(n).
+    /// </summary>
+    public class KmpSearcher
+    {
+        /// <summary>
+        /// Enumerates each zero-based index of all occurrences of the string in this instance.
+        /// throws ItemNotFoundException if the item couldn't be found.
+        /// </summary>
+        /// <param name="str">The string instance.</param>
+        /// <param name="pat">The pattern to seek.</param>
+        /// <returns>
+        /// The zero-based index positions of value if one or more <paramref name="pat"/> are found.
+        /// If <paramref name="pat"/> is empty, no indexes will be enumerated.
+        /// </returns>
+        /// <exception cref="ItemNotFoundException"><paramref name="str"/> or <paramref name="pat"/> is null.</exception>
+        public IEnumerable<int> FindIndexes(string str, string pat)
+        {
+            List<int> retVal = new List<int>();
```

Comment:
Could you make this a generator function, I think it could be useful taking into account that IEnumerable is returned? You can leave it as it is if requested changes are difficult to make or make code less readable or something else.

Reply:
It's possible but a little costly. 
If there was no `ItemNotFoundException `I could replace the `return `with `yield return`.
Anyway, since [`List<T>` implements the `IEnumerable<T>`](https://stackoverflow.com/a/4700629/579381) , I change return type to List<int>.